README.md
# Kasm Workspaces

[Kasm Workspaces](https://kasmweb.com) provides enterprise-class orchestration,
data loss prevention, and web streaming technology to enable the delivery of
containerized workloads to your browser and desktops to end-users.

## Table of Contents

- [Installation](#installation)
  - [Download and extract](#download-and-extract)
  - [To run Kasm on port 8443 with a reverse proxy on 443](#to-run-kasm-on-port-8443-with-a-reverse-proxy-on-443)
- [Custom Docker Images](#custom-docker-images)
- [Cron periodic cleanup and certificate renewal](cron/README.md)
- [References](#references)

## Installation

### Download and extract

```bash
[ -d $HOME/Kasm ] || mkdir -p $HOME/Kasm
cd $HOME/Kasm
curl -O https://kasm-static-content.s3.amazonaws.com/kasm_release_1.14.0.3a7abb.tar.gz
tar -xf kasm_release_1.14.0.3a7abb.tar.gz
sudo bash kasm_release/install.sh --accept-eula --swap-size 8192
```

### To run Kasm on port 8443 with a reverse proxy on 443

```bash
sudo bash kasm_release/install.sh -L 8443 --accept-eula --swap-size 8192
```

- Log into the Web Application at https://<WEBAPP_SERVER>
- Default usernames are `admin@kasm.local` and `user@kasm.local`
- Passwords will be randomly generated and presented at the end of the install
  - Unless the `--admin-password` and/or `--user-password` are specified

## Custom Docker Images

See the Kasm Workspaces documentation on
[Building Custom Images](https://kasmweb.com/docs/latest/how_to/building_images.html#).

Many example custom Kasm image builder Dockerfiles can be found at
https://github.com/doctorfree/workspaces-images including a `Neovim` image.
To build the `Neovim` Workspace image:

```bash
# On the Kasm Workspaces server
git clone https://github.com/doctorfree/workspaces-images
cd workspaces-images
export GH_TOKEN="<YOUR-GITHUB-API-TOKEN>"
./build-neovim
```

After building a custom image, as an administrator in Kasm Workspaces,
add the custom Workspace with `Workspaces -> Workspaces -> Add Workspace`.
Add the `Neovim` image as:

- **Workspace Type**: `Container`
- **Thumbnail URL**: `https://lazyman.dev/assets/neovim-hicontrast.png`
- **Docker Image**: `kasmweb/neovim:dev`
- **Cores**: `2`
- **Memory**: `5636`
- **GPU Count**: `0`
- **CPU Allocation Method**: `inherit`
- **Persistent Profile Path**: `/u/kasm_profiles/{image_id}/{user_id}`
- **Docker Run Config Override**: `{ "hostname": "kasm" }`

After adding a Workspace, check `Workspaces -> Registry -> Installed Workspaces`
to verify the Workspace is installed. The first time this Workspace session runs
the [Lazyman Neovim Configuration Manager](https://github.com/doctorfree/nvim-lazyman)
is installed and configured.

## References

- [Kasm Documentation](https://www.kasmweb.com/docs/latest/index.html)
- [Kasm Administration](https://www.kasmweb.com/docs/latest/admin_guide.html)
- [Kasm User Guide](https://www.kasmweb.com/docs/latest/user_guide.html)
- [Kasm Workspaces - Resource Allocation](https://youtu.be/lv85XZ8EdjY?si=xcSfB-EWtu-2tIHQ) (YouTube)
- [KasmVNC](https://www.kasmweb.com/kasmvnc)
- [Record Technologies Kasm Registry](https://doctorfree.github.io/kasm-registry/1.0)
  - [Github Repository](https://github.com/doctorfree/kasm-registry) (deployment details)
- [LinuxServer.io](https://www.linuxserver.io)
  - [3rd Party Workspace Registry](https://kasmregistry.linuxserver.io)
  - [Generate Add Workspace JSON](https://kasmregistry.linuxserver.io/1.0/new)
- [Dockerizing a node.js Web Application](https://semaphoreci.com/community/tutorials/dockerizing-a-node-js-web-application)


ReleaseNotes.md
![Logo](https://raw.githubusercontent.com/wiki/doctorfree/kasm-registry/logo.png)

# Record Technologies Kasm Workspace Images

The [Record Technologies Kasm repository](https://doctorfree.github.io/kasm-registry) contains several highly customized desktop and application Kasm Workspaces images. Administrators may leverage these images directly or use them as a starting point for their own custom images. Each of these images is based off one of the [**Workspaces Core Images**](https://github.com/kasmtech/workspaces-core-images) which contain the necessary wiring to work within the Kasm Workspaces platform.

For more information about building custom images please review the [**How To Guide**](https://kasmweb.com/docs/latest/how_to/building_images.html)

The Kasm team publishes applications and desktop images for use inside the platform. More information, including source can be found in the [**Default Images List**](https://kasmweb.com/docs/latest/guide/custom_images.html)

## Table of Contents

- [Record Technologies Workspace Registry](#record-technologies-workspace-registry)
  - [Workspace features](#workspace-features)
  - [Registry deployment](#registry-deployment)
- [Repository Structure](#repository-structure)
- [Manual Deployment](#manual-deployment)
  - [Manual Execution](#manual-execution)
- [About Workspaces](#about-workspaces)
- [Live Demo](#live-demo)

## Record Technologies Workspace Registry

![Registry](https://raw.githubusercontent.com/wiki/doctorfree/kasm-registry/registry.png)

The [Record Technologies](https://doctorfree.github.io/kasm-registry) workspace registry serves as the distribution mechanism for Kasm workspaces generated from [Doctorfree Open Source Projects](https://github.com/doctorfree).

Currently available workspaces in this registry include:

* `AppImage`: Customized Ubuntu 22.04 desktop with [AppImage Launcher](https://github.com/TheAssassin/AppImageLauncher)
* [Asciiville](https://github.com/doctorfree/Asciiville#readme): Ascii art, utilities, games, more
* [Ascii Games](https://github.com/doctorfree/asciigames#readme): Collection of Ascii games
* `CloudStorage`: Utilizes [rclone](https://github.com/rclone/rclone), a command-line program to manage files on cloud storage
* `Deluxe`: Asciiville, Neovim, Spiderfoot, Wing and more all in one workspace
* [LM Studio](https://github.com/lmstudio-ai): Customized Ubuntu 22.04 desktop with `LM Studio`
* [Neovim](https://github.com/doctorfree/nvim-lazyman#readme): Neovim, neovide, lazyman, much more
* [Open-Source Intelligence](https://en.wikipedia.org/wiki/Open-source_intelligence): Reconnaissance tool, forensics, link analyzer, ...
* [Project Jupyter](https://en.wikipedia.org/wiki/Project_Jupyter): Jupyter notebooks
* [Spiderfoot](https://github.com/smicallef/spiderfoot): A reconnaissance tool that automatically queries public data sources to gather intelligence
* [Ubuntu desktops](https://en.wikipedia.org/wiki/Ubuntu): Customized Ubuntu Focal and Jammy desktops
* [Wing cloud programming language](https://www.winglang.io): Wing programming language, examples, and editors
* `WingPlus`: The Wing workspace with the Neovim hyper-extensible text editor
* More to come ...

### Workspace features

Record Technologies workspaces all include customized desktops with several productivity and development packages preconfigured for ease of use. In addition to the rich feature set provided by the base images upon which these workspaces are built, Record Technologies workspaces provide advanced configuration and integration of many additional components including:

* [Bat](https://github.com/sharkdp/bat): A `cat(1)` clone with wings
* [Btop](https://github.com/aristocratos/btop): Resource monitor
* [Cheat Sheets](https://github.com/doctorfree/cheat-sheets-plus): Cheat sheets Obsidian vault
  - Integrated with `ranger` and `glow` for terminal/console viewing
* [Fzf](https://github.com/junegunn/fzf): A command-line fuzzy finder
* [Glow](https://github.com/charmbracelet/glow): Markdown previewer
* [Go](https://go.dev): A modern version of the Go programming language is installed
* [Kitty](https://sw.kovidgoyal.net/kitty): The fast, feature-rich, GPU based terminal emulator
* [Lsd](https://github.com/lsd-rs/lsd): The next generation `ls` command
* [Neofetch](https://github.com/dylanaraps/neofetch): System information tool
* [Node](https://nodejs.org): Node version 20
  * [Nvm](https://github.com/nvm-sh/nvm): Node version manager
* [Nerd Fonts](https://www.nerdfonts.com): Large number of extra glyphs and icons
* [Obsidian](https://obsidian.md): Markdown based personal knowledge base and note-taking application
* [Ranger](https://github.com/ranger/ranger): A Vim-inspired file manager for the terminal
  - Extensive configuration for the Ranger file executor/opener `rifle`
  - Integration with `Glow`, `AppImage`, `Wing`, and more for execution/opening of selections
* [Vim](https://www.vim.org): Preconfigured `vim` text editor with plugins, keymaps, theme, etc
* [Xfce4](https://www.xfce.org): Customized Xfce desktop environment
* [Zoxide](https://github.com/ajeetdsouza/zoxide): Smart `cd` command
* [Zsh](https://www.zsh.org): Features from `bash`, `ksh`, `tcsh`, and more
  * [Oh My Zsh](https://ohmyz.sh): Framework for managing the `zsh` configuration
  * [Powerlevel10k](https://github.com/romkatv/powerlevel10k): Theme for `zsh`

The Wing workspaces include Visual Studio Code with the Wing extension; the Neovim workspaces include Neovide and Lazyman - the Neovim Configuration Manager, several workspaces include the [Ranger File Manager](https://github.com/ranger/ranger) with customized launch configurations and the [Btop++ system monitor](https://github.com/aristocratos/btop) with customized configuration and theme.


web/reinstall.sh
#!/bin/bash
#
HERE="$( cd -- "$(dirname "$0")" >/dev/null 2>&1 ; pwd -P )"
TOP=$(dirname "${HERE}")

export DEBIAN_FRONTEND=noninteractive

sudo systemctl stop nginx

# Install dependencies and tools
sudo apt remove nginx -q -y
sudo apt autoremove -q -y
sudo apt install nginx -q -y

[ -d /etc/nginx/sites-available ] || sudo mkdir -p /etc/nginx/sites-available
[ -f kasm01.hornblower.com ] && {
    sudo cp kasm01.hornblower.com /etc/nginx/sites-available
    sudo rm -f /etc/nginx/sites-enabled/kasm01.hornblower.com
    sudo ln -s /etc/nginx/sites-available/kasm01.hornblower.com \
               /etc/nginx/sites-enabled/kasm01.hornblower.com
    sudo rm -f /etc/nginx/sites-enabled/default
}
[ -d /etc/nginx/ssl ] || sudo mkdir -p /etc/nginx/ssl
[ -f ${TOP}/certs/wildcard/server.crt ] && {
    sudo cp ${TOP}/certs/wildcard/server.crt /etc/nginx/ssl
    sudo chmod 444 /etc/nginx/ssl/server.crt
}
[ -f ${TOP}/certs/wildcard/server.key ] && {
    sudo cp ${TOP}/certs/wildcard/server.key /etc/nginx/ssl
    sudo chown root /etc/nginx/ssl/server.key
    sudo chmod 400 /etc/nginx/ssl/server.key
}
[ -d /etc/nginx/snippets ] || sudo mkdir -p /etc/nginx/snippets
[ -f letsencrypt.conf ] && sudo cp letsencrypt.conf /etc/nginx/snippets
[ -f ssl.conf ] && sudo cp ssl.conf /etc/nginx/snippets

sudo systemctl daemon-reload
sudo systemctl enable --now nginx
sudo systemctl start nginx


web/install.sh
#!/bin/bash
#
HERE="$( cd -- "$(dirname "$0")" >/dev/null 2>&1 ; pwd -P )"
TOP=$(dirname "${HERE}")

export DEBIAN_FRONTEND=noninteractive

sudo apt install nginx -q -y

[ -d /etc/nginx/sites-available ] || sudo mkdir -p /etc/nginx/sites-available
[ -f kasm01.hornblower.com ] && {
  sudo cp kasm01.hornblower.com /etc/nginx/sites-available
  sudo rm -f /etc/nginx/sites-enabled/kasm01.hornblower.com
  sudo ln -s /etc/nginx/sites-available/kasm01.hornblower.com \
             /etc/nginx/sites-enabled/kasm01.hornblower.com
  sudo rm -f /etc/nginx/sites-enabled/default
}
[ -d /etc/nginx/ssl ] || sudo mkdir -p /etc/nginx/ssl
[ -f ${TOP}/certs/wildcard/server.crt ] && {
  sudo cp ${TOP}/certs/wildcard/server.crt /etc/nginx/ssl
  sudo chmod 444 /etc/nginx/ssl/server.crt
}
[ -f ${TOP}/certs/wildcard/server.key ] && {
  sudo cp ${TOP}/certs/wildcard/server.key /etc/nginx/ssl
  sudo chown root /etc/nginx/ssl/server.key
  sudo chmod 400 /etc/nginx/ssl/server.key
}
[ -d /etc/nginx/snippets ] || sudo mkdir -p /etc/nginx/snippets
[ -f letsencrypt.conf ] && sudo cp letsencrypt.conf /etc/nginx/snippets
[ -f ssl.conf ] && sudo cp ssl.conf /etc/nginx/snippets

for ngc in nginx/*
do
  cfg=$(basename ${ngc})
  [ -f /etc/nginx/${cfg} ] || sudo cp ${ngc} /etc/nginx/${cfg}
done

sudo systemctl --quiet restart nginx

# Install dependencies and tools
sudo apt install fail2ban python3-psycopg2 \
                 nginx certbot python3-certbot-nginx -q -y

sudo openssl dhparam -out /etc/ssl/certs/dhparam.pem 2048 > /dev/null 2>&1
sudo mkdir -p /var/lib/letsencrypt/.well-known
sudo chgrp www-data /var/lib/letsencrypt
sudo chmod g+s /var/lib/letsencrypt
sudo systemctl --quiet restart nginx
# We are using a wildcard certificate for now
# Let's Encrypt tries to perform a challenge but fails due to our firewall
sudo certbot certonly --agree-tos --email ron.record@hornblower.com \
             --webroot -w /var/lib/letsencrypt/ -d kasm01.hornblower.com
sudo systemctl --quiet reload nginx

echo "# Restart nginx when certificate is renewed" > /tmp/cli$$
echo "deploy-hook = systemctl reload nginx" >> /tmp/cli$$
if [ -f /etc/letsencrypt/cli.ini ]; then
  grep "deploy-hook" /etc/letsencrypt/cli.ini > /dev/null || {
    sudo cat /etc/letsencrypt/cli.ini /tmp/cli$$ > /tmp/lcl$$
    sudo cp /tmp/lcl$$ /etc/letsencrypt/cli.ini
  }
else
  sudo cp /tmp/cli$$ /etc/letsencrypt/cli.ini
fi
rm -f /tmp/cli$$

[ -f /etc/fail2ban/filter.d/send.conf ] || {
  [ -d /etc/fail2ban/filter.d ] || sudo mkdir -p /etc/fail2ban/filter.d
  sudo cp f2b-send.conf /etc/fail2ban/filter.d/send.conf
}
[ -f /etc/fail2ban/jail.d/jail.local ] || {
  [ -d /etc/fail2ban/jail.d ] || sudo mkdir -p /etc/fail2ban/jail.d
  sudo cp f2b-send.local /etc/fail2ban/jail.d/jail.local
}

sudo systemctl --quiet daemon-reload
sudo systemctl --quiet enable --now nginx
